### DISCUSSION TITLE
Creating a Single Page Application (SPA) with NiceGUI

### DISCUSSION THREAD

**Discussion by Diegiwg** [zauberzeug/nicegui#349] [Show and tell]
https://user-images.githubusercontent.com/15692310/217125810-af5e39c3-70cc-4779-9d95-3f4f60228e33.mp4

- I would like to update only a certain part of my application, keeping the menu layout intact, as it is always the same on all 'pages'.
- For that, I tried to find a method to force the full rendering of a specific component, but I didn't find anything in the NiceGUI API.
- Then, looking a little further, I understood where the 'elements' are stored, and how I could transfer the responsibility of these arguments to another Element.

Using that as a base, I created the following module:

`routes.py`
```python
from dataclasses import dataclass
from typing import Callable
from nicegui.element import Element


@dataclass
class Route:
    path: str
    title: str
    handler: Callable


class RouteManager:
    routes: dict[str, Route]
    current_route: Route

    def __init__(self) -> None:
        self.routes = dict()

    def register_route(self, route: Route):
        self.routes[route.path] = route

    def set_route(self, route_path: str):
        self.current_route = self.routes[route_path]

    async def run_route(self):
        return await self.current_route.handler()


route_manager = RouteManager()


@dataclass
class State:
    root: Element
    element: Element


state = State(root=Element(""), element=Element(""))


@dataclass
class Redirect:
    route_path: str

    async def run(self):
        route_manager.set_route(self.route_path)
        temp: Element = await route_manager.run_route()

        state.element.clear()
        state.element.slots = temp.slots
        state.root.remove(temp)

        state.element.style(add="width: 100%; height: 95vh;")

```

Note: it will always report an error when changing pages, because the Temporary Element no longer exists, but the rendering loop will still try to render it. This does not break anything in the application.

To use the system, just define some pages, like the examples below:

`sales.py`
```python
from nicegui import ui
from nicegui.page import globals

import components
import routes


async def page():
    with ui.column().style(add="width: 100%; height: 95vh;") as element:
        globals.title = routes.route_manager.current_route.title
        components.page_title(routes.route_manager.current_route.title)

        ui.label("WIP").classes(add="text-center w-full")

    return element


sales_route = routes.Route("sales", "Vendas", page)
routes.route_manager.register_route(sales_route)
```

Note: always remember to structure the first element as ui.column, as it is the element with the closest behavior to the base @ui.page.

And in the main of your application, define the root page “/”, with the following initial structure:

`main.py`
```python
from nicegui import Client, app, ui

import components
import pages
import routes

routes.route_manager.set_route("sales")


@ui.page("/")
async def render(client: Client):
    with ui.card().style(add="width: 100%; height: 95vh;"):
        client.on_connect(components.render_layout)

        with ui.column().style(add="width: 100%; height: 95vh;") as root:
            element = await routes.route_manager.run_route()

        routes.state.root = root
        routes.state.element = element


app.add_static_files("/static", "static")
ui.run(binding_refresh_interval=0.5, uvicorn_logging_level="")
```

That's it, now with buttons like the ones below, you can easily navigate between your pages, updating only the necessary components:

`layout.py`
```python
from nicegui import ui

from routes import Redirect
from components import clock

def menu_button(text: str, path: str):
    return ui.button(text, on_click=Redirect(path).run).style(add="width: 100%;")


async def render():
    with ui.left_drawer():
        with ui.column():
            with ui.row().style(
                add="display: flex; align-items: center; justify-content: center; width: 100%;"
            ):
                ui.image(source="/static/logo.png").props(add="width='50px'")
                ui.label("QUEIROZ LUBRIFICANTES").style(add="")
            ui.separator()

            with ui.column().style(
                add="width: 100%; height: 85vh; display: grid; align-content: space-between;"
            ):
                with ui.column().style(add="width: 100%;"):
                    menu_button("Vendas", "sales")
                    menu_button("Produtos", "products")
                    menu_button("Estoque", "stock")

                await clock()

```

> It's not perfect, but it gives an implementation idea for rendering all children of a component via update.
> Remembering, I don't know if this update system exists in the NiceGUI API. If there is, let me know.

**Comment by rodja** [comment]
Very nice. Build-in SPA support is surely a good thing for NiceGUI in the long run. Your code shows a first implementation. I like the introduction of `Route` and `RouteManager` as it adds some tooling support. Whenever we need SPA like functionality we currently do it this way:

```py
from nicegui import ui

def show_one():
    content.clear()
    with content:
        ui.label('Content One').classes('text-2xl')


def show_two():
    content.clear()
    with content:
        ui.label('Content Two').classes('text-2xl')


def show_three():
    content.clear()
    with content:
        ui.label('Content Three').classes('text-2xl')


with ui.row():
    ui.button('One', on_click=show_one)
    ui.button('Two', on_click=show_two)
    ui.button('Three', on_click=show_three)

content = ui.column()
show_one()

ui.run()
```

While it works nicely and is also quite short, it has by no means any framework to support history, location path updates and other things which can easily be integrated in your solution. Maybe the build-in support can be based on your code? Another approach would be to look at [Vue.js router](https://router.vuejs.org/guide/#javascript) and use that as foundation...

**Comment by Diegiwg** [comment]
Damn, I didn't know you could do `with Elementor` and get the context of the element.

I used it as a base here, and I rewrote the Redirect.run function.

No more errors now, thank you.

[Link to GIST with complete example code](https://gist.github.com/Diegiwg/0b73cf36c4b3146a85d28e7dd8388fcb)

**Reply by Diegiwg** [reply]
Example Code:

```python
from dataclasses import dataclass
from typing import Callable
from nicegui.element import Element
from nicegui import ui, Client


#### ROUTES.PY ####
@dataclass
class Route:
    path: str
    title: str
    handler: Callable


class RouteManager:
    routes: dict[str, Route]
    current_route: Route

    def __init__(self) -> None:
        self.routes = dict()

    def register_route(self, route: Route):
        self.routes[route.path] = route

    def set_route(self, route_path: str):
        self.current_route = self.routes[route_path]

    async def run_route(self):
        await self.current_route.handler()


route_manager = RouteManager()


@dataclass
class State:
    element: Element


state = State(element=Element(""))


@dataclass
class Redirect:
    route_path: str

    async def run(self):
        route_manager.set_route(self.route_path)
        return await route_manager.run_route()


#### LAYOUT.PY ####
def menu_button(text: str, path: str):
    return ui.button(text, on_click=Redirect(path).run).style(add="width: 100%;")


async def layout_render():
    with ui.left_drawer():
        with ui.column():
            with ui.row().style(
                add="display: flex; align-items: center; justify-content: center; width: 100%;"
            ):
                ui.label("MENU")
            ui.separator()

            with ui.column().style(
                add="width: 100%; height: 85vh; display: grid; align-content: space-between;"
            ):
                with ui.column().style(add="width: 100%;"):
                    menu_button("Page 01", "page-1")
                    menu_button("Page 02", "page-2")


#### PAGE_01.PY ####
async def page_01():
    state.element.clear()
    with state.element:
        ui.label("Page 01")


page_01_route = Route("page-1", "Page 01", page_01)
route_manager.register_route(page_01_route)


#### PAGE_02.PY ####
async def page_02():
    state.element.clear()
    with state.element:
        ui.label("Page 02")


page_02_route = Route("page-2", "Page 02", page_02)
route_manager.register_route(page_02_route)

#### MAIN.PY ####
route_manager.set_route("page-1")


@ui.page("/")
async def render(client: Client):
    client.on_connect(layout_render)
    state.element = ui.card().style(add="width: 100%; height: 95vh; background: grey;")
    await route_manager.run_route()


ui.run(dark=True)
```

**Reply by rodja** [reply]
Nice!

**Comment by rodja** [comment]
We just merged an very simple SPA example into main: https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app

**Reply by ghost** [reply]
This SPA example is very clever, but  I think it's limited by popstate with the URL-liked SPA route string, with this method, there is only one RouteFrame object exist which is the container for context,  I can't create more than one RouteFrame object to split the context into multi sub-contexts just bucause the URL-liked SPA route ONLY one can appear to maintain the brower history record, and this method causes a conflict between SPA Router(javascript event trigger) and FastAPI‘s Router(HTTP trigger) because both use URL string.  So how about use another syntax for SPA route which different from HTTP URL? the advatages about the new syntax SPA route are: 1) can free split the context into sub-contexts to reflesh sub-contexts Independently or reflesh the whole context because the SPA route is internal used no matter consider the browser history record (if needed, there can do a translation in JavaScript event). 2) Can be mixed SPA with MPA.

**Reply by rodja** [reply]
Sure. It's only an example. And we are still not sure how the API should be. Maybe your points should be discussed in  https://github.com/zauberzeug/nicegui/discussions/1066...

**Comment by aisha0801** [comment]
Hello, I'm relatively new to coding and NiceGUI, and I'm attempting to learn by dissecting your code. I've replicated the files you've provided here, but upon running the code, I encountered an error: ModuleNotFoundError: No module named 'components'. Could someone please guide me on how to troubleshoot this issue? Is 'components' supposed to be another module, or is it something that needs to be installed via the terminal?

**Reply by rodja** [reply]
Have you tried https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app? The code directly provided in this old discussion may not work anymore due to changes made over the time in NiceGUI. We only update the official examples.

**Comment by rotrigoh** [comment]
Ficou muito massa, mano! Parabéns pelo trabalho. Sabes me dizer se você ainda continua utilizando o nicegui para criar essas ferramentas? Te atende bem?

**Reply by falkoschindler** [reply]
@rotrigoh Please use English language when posting to this repository, e.g. by using an online translation tool. Otherwise we need to do the translation ourselves and can't be sure the result is what you meant.

**Reply by rotrigoh** [reply]
Sorry... I ended up realizing that our friend in the topic was Brazilian, so I decided to speak in our language. Anyway, I'll pay attention next time.

**Comment by v479038280** [comment]
beautiful example ,

**Comment by rodja** [comment]
Just for reference: in #2811 @alyxion created a beautiful pull-request to provide direct API support for SPAs in NiceGUI. Hopefully it will be merged in the next few weeks.

**Reply by Alyxion** [reply]
Definitely. Was the last weeks on the road due to work w/ quite long days and if you have at 9 in the evening the choice between hotel spa :sauna_man:  and evening beer :beer: or working on a 13 inch laptop :notebook: in the hotel room - choice is easy - but will be back on Monday and continue where I stopped. PR should latest be finished by mid May, so stay tuned. :sunglasses:

**Reply by me21** [reply]
What's the difference between SPA and regular NiceGUI workflow when you just remove some elements from the page and create others?

**Reply by falkoschindler** [reply]
See https://en.wikipedia.org/wiki/Single-page_application. With the mentioned pull request we try to find an API for cleverly adding and removing elements so that it feels like navigating the app, even though the browser tab never has to reload the page. Of course, you can add and remove elements yourself, but for a large application it is useful to hide away the details.

**Comment by deveapi** [comment]
Is it possible to make the navbar items look as the usual menu items list instead of button? I tried `ui.link` but it does not support `router.open`
Currently
```
	# Navigation bar
	with ui.header().classes('bg-blue-500 text-white p-4'):
		with ui.row().classes('items-center w-full'):
			with ui.row().classes('gap-6'):
				ui.button('Home', on_click=lambda: router.open('/')).classes(
					'bg-transparent text-white text-left text-lg px-4 py-2 hover:bg-blue-600 cursor-pointer'
				)
				ui.button('Prompt', on_click=lambda: router.open('/prompt')).classes(
					'bg-transparent text-white text-left text-lg px-4 py-2 hover:bg-blue-600 cursor-pointer'
				)
```
![image](https://github.com/user-attachments/assets/5ee43562-1149-4787-93eb-f6c9ac467bef)
For example wanna to be like this and also potentially sub menu items for more functionality
![image](https://github.com/user-attachments/assets/41e557a5-85fa-4b34-9423-a48b298fb705)

**Reply by Alyxion** [reply]
@deapi, as @rodja already mentioned, there is already a pull request in progress which will make this possible.

In that #2811 he SPA is a really integrated in NiceGUI and knows which links are part of the SPA and which ones are not, so ui.link works as you intended for use case. 

See https://github.com/zauberzeug/nicegui/pull/2811/files#diff-1b2b4e6c42c15b832e79c4528bd801ad42660d96b4965e001966c87a3fbc07a5

**Comment by ravi-2912** [comment]
Hi all, 
Excellent work and love the NiceGUI and being far better than other framework I am now shifting my apps to NiceGUI, however I am still young in NiceGUI.
I am using the SPA example from the NiceGUI webiste and here is my modified version, [https://github.com/ravi-2912/nicegui-spa](https://github.com/ravi-2912/nicegui-spa)
Here is a video demo

https://github.com/user-attachments/assets/4bd33e05-6dc9-422a-9728-995c8a3ff1a1

Although it is going fine, there are some challenges I am facing and hope the community could guide.
1. When changing the path directly in the address bar, I don't see error pages, but the content of the SPA page goes black. How can I get error page?
2.  Likewise when raising an exception or manually adding an exception in the code does not show error pages.
3. I understand that the drawers are top level elements, however I intend to use right drawer to render per page custom UI as show in video.  I am able to achieve this by storing the `ui.right_drawer` instance in `app.storage.client` however I don't like this approach and moreover I have to clean the content of the right_drawer instance on every page before I can render the right one. Any suggestions?
4. Also I want to allow user to change the right drawer width using a slider. Any recommendations how can this be done?

**Comment by rodja** [comment]
All good questions @ravi-2912! We are currently hard at work to get full-fledged SPA support in #4821.

> When changing the path directly in the address bar, I don't see error pages, but the content of the SPA page goes black. How can I get error page?

Will be fixed with #4821.

> Likewise when raising an exception or manually adding an exception in the code does not show error pages.

Is also part of #4821.

> storing the ui.right_drawer instance in app.storage.client

#4821 will have build-in data passing.

> moreover I have to clean the content of the right_drawer instance on every page

Thats correct and will not change. Calling clear on an element and then add the new content is ok. The `ui.sub_pages` and `ui.refreshable` also do this under the hood.

> Also I want to allow user to change the right drawer width using a slider.

This is best run as a separate question because it is not specific to SPAs.

**Reply by ravi-2912** [reply]
Thanks @rodja
